### DISCUSSION TITLE
Passing Enums to Unicorn views

### DISCUSSION THREAD

**Discussion by nerdoc** [django-commons/django-unicorn#345] [Ideas]
According to the [docs](https://www.django-unicorn.com/docs/actions/#argument-types), passed types must be simple Python types.
I am struggling passing an IntEnum, which in fact is an int, to a view:

```django
<button 
  type="button" 
  class="..." 
  unicorn:click="set_scope({{ scope.DEVICE }})"
>
set scope to DEVICE
</button>
```
But I get no way to solve this. can I pass the Enum ("Scope" here) as attribute to the view?

```python 
class FooView(UnicornView):
    scope = Scope.USER
  # ...
```
And ideas? Would it be possible to cast Enums/IntEnums to str or ints in Unicorn? I always get the error:
```
ValueError: malformed node or string on line 1: <ast.Attribute object at 0x7f83bcbbdae0>
```

**Comment by nerdoc** [comment]
Hm... I found a bit of an answer after hours now. You have to keep the variable as Enum in the View, and cast it to an Enum when getting it from the template.

```python
class ScopeSwitcherView(UnicornView):
    active_scope: Scope = Scope.USER

    def set_scope(self, scope: int):
        self.active_scope = Scope(scope)
```
In the template, use ints/strings, by using `{{ <enum_name>.<ATTR>.value }}`:
```django

<button
    type="button"
    class="btn {% if active_scope == active_scope.GROUP.value %} active{% endif %}"
    unicorn:click="set_scope({{ active_scope.GROUP.value }})"
>Set to GROUP</i></button>
```

**Reply by adamghill** [reply]
Glad you found a workaround -- it makes sense, but let me see if I can write a test to see if supporting enums directly would be possible.

**Reply by adamghill** [reply]
I was trying to see if I had enough context when parsing the method call name (i.e. `set_scope(active_scope.USER)`) to automatically deal with the enum for you, but it looks like I don't. In other words, I can see `active_scope.USER` but I don't have any concept of where that `active_scope` is defined or the underlying enum of `Scope`.

In normal Django templates I've always had to use `.value` on the enum to get the underlying int/str, so I guess this makes sense? I can add a note to the docs as a heads up for others though in case they run into the same situation.

**Reply by nerdoc** [reply]
That would be great. to sum it up: keep values in the template as simple types, and cast it to e.g. an Enum when entering the component in the backend.

**Reply by adamghill** [reply]
I added a note about enums here: https://www.django-unicorn.com/docs/actions/#argument-types. Let me know if it could be clarified. I also added an example here: https://github.com/adamghill/django-unicorn/blob/main/example/unicorn/templates/unicorn/objects.html#L108-L117.

**Reply by nerdoc** [reply]
Very clear. Maybe I would add a comment, and move this one at the last position.

#### first
I would use `set_color({{ color.BLUE.value }})"` as first one. It's the clearest...

#### second
I would add a comment and use that as second (= last one)
```html
<!-- the following is not recommended, as it contradicts the purpose of Enums (but it works): -->
  <button unicorn:click="set_color(2)">Sets `self.color` to `Color.GREEN`</button> 
```
#### third...
`set_color({{ color.value }})"` - eh - does this do anything? because `color.value` is already `color.value`... This would never change anything... Couldn't you delete that example altogether?

**Reply by adamghill** [reply]
>because color.value is already color.value... This would never change anything

Yup! I think what I was trying to make clear was an enum could be used directly (if that makes sense). Does this clarify what I was getting at? It still might not be all that helpful. 😆 

```python
# enum.py
from django_unicorn.components import UnicornView
from enum import Enum

class Color(Enum):
  RED = 1
  GREEN = 2
  BLUE = 3
  PURPLE = 4

class EnumView(UnicornView):
    color = Color.RED
    purple_color = Color.PURPLE

    def set_color(self, color_int: int):
      self.color = Color(color_int)
```

```html
<!-- enum.html -->
<div>
  <button unicorn:click="set_color({{ color.BLUE.value }})">Will output BLUE below</button>
  <button unicorn:click="set_color(2)">Will output GREEN below</button>
  <button unicorn:click="set_color({{ purple_color.value }})">Will output PURPLE below</button>

  <br />
  Color: {{ color }}
</div>
```

**Reply by nerdoc** [reply]
MUCH better!

**Reply by adamghill** [reply]
Commit is https://github.com/adamghill/django-unicorn/commit/dd726d9c2988276ebd681b010970f5f0a303e9e1; should go live to the site in a little bit.

Thanks for your help!
